server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -328,7 +326,7 @@ public static Builder builder(Index index) {
     public static class Builder {
 
         private final Index index;
-        private final Map<Integer, IndexShardRoutingTable> shards = new HashMap<>();
+        private IndexShardRoutingTable.Builder[] shards;
```

Comment:
I think the primary reason we use an array rather than a List here is to size it properly, in particular for the 1 shard case, perhaps we should comment on that?

Reply:
I am not sure I follow. You have `ensureShardArray` anyway, which could be translated directly into `ArrayList.ensureCapacity`? It could avoid the potential degenerate case of copying data over and over if growing by one too many times.

But let us leave this out, was a minor point for sure.

Reply:
> You have ensureShardArray anyway, which could be translated directly into ArrayList.ensureCapacity?

It doesn't translate like that. `ensureCapacity` still doesn't change the list size. Even if I do `ensureCapacity(2)` I don't automatically get the ability to add at index `1`, I still need to put a `null` into index `0` to be able to do that without an index out of bounds exception :)